utils/refchecker.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
+
+import sys
+import os
+import argparse
+import json
+
+import ssg.build_yaml
+import ssg.products
+import ssg.rules
+import ssg.yaml
+import ssg.utils
+import ssg.rule_yaml
+
+SSG_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Check all rule.yml referenced in a given"
+                                     "profile for a required reference identifier")
+    parser.add_argument("-j", "--json", type=str, action="store",
+                        default="build/rule_dirs.json", help="File to read "
+                        "json output of rule_dir_json from (defaults to "
+                        "build/rule_dirs.json")
+    parser.add_argument("-c", "--build-config-yaml", default="build/build_config.yml",
+                        help="YAML file with information about the build configuration. "
+                        "Defaults to build/build_config.yml")
+    parser.add_argument("-p", "--profiles-root",
+                        help="Override where to look for profile files.")
+    parser.add_argument("product", type=str, help="Product to check has required references")
+    parser.add_argument("profile", type=str, help="Profile to iterate over")
+    parser.add_argument("reference", type=str, help="Required reference system to check for")
+
+    return parser.parse_args()
+
+
+def load(rule_obj, env_yaml=None):
+    """
+    From the given rule_dir object, load the build_yaml.Rule associated with
+    it.
+    """
+
+    yaml_file = ssg.rules.get_rule_dir_yaml(rule_obj['dir'])
+    return ssg.build_yaml.Rule.from_yaml(yaml_file, env_yaml=env_yaml)
+
+
+def load_for_product(rule_obj, product, env_yaml=None):
+    """
+    From the given rule_dir object, load the build_yaml.Rule associated with
+    it, normalizing for the given product.
+    """
+
+    rule = load(rule_obj, env_yaml=env_yaml)
+    rule.normalize(product)
+    return rule
+
+
+def reference_check(env_yaml, rule_dirs, profile_path, product, reference):
+    profile = ssg.build_yaml.ProfileWithInlinePolicies.from_yaml(profile_path, env_yaml)
```

Comment:
I have reviewed it again and I have found one more thing. There's a situation in which it doesn't work form me. I have removed the anssi reference from rule selinux_state and then I tried if the tool reports it, but it doesn't. ` utils/refchecker.py  rhel8 anssi_bp28_high anssi` gives no output. I think it's because of the controls feature used in anssi profiles. I think you need to run `resolve` method on the `profile`.

Reply:
Well that was a bit of work ;-)

I ended up refactoring `compile_profiles.py` to make this possible. Let me know what you think.